File: 1 Python/solutions/sock_sorter.py
# sock_sorter.py
"""
Hints:
- Loop through 100
- Use a dictionary as a counter for your socks
- At the end, use a little math to figure out if your 
  sock count is even 
"""

# counter = {}

# # creating a count 
# counter['thing'] = 1 

# # increasing count 
# counter['thing'] += 1

# # or, in one line,
# # if count['thing'] exists, increase its count by 1
# # else, set its count as 1
# counter['thing'] = counter.get('thing', 0) + 1

import random

sock_list = []
sock_types = ['ankle', 'crew', 'calf', 'thigh']
colors = ['black', 'white', 'blue']
sock_counter = {}

for i in range(100):
    sock = random.choice(sock_types)
    color = random.choice(colors)
    sock_list.append((sock, color))
    
    sock_counter[(sock, color)] = sock_counter.get((sock, color), 0) + 1
# 
#     # equivalent to above
#     # if sock in sock_counter:
#     #     sock_counter[sock] += 1 
#     # else:
#     #     sock_counter[sock] = 1

print(sock_list)
print(sock_counter)

for sock in sock_counter:
    print(f'{sock} has {sock_counter[sock]%2} loner(s)')
    # print(f'{sock} has {sock_counter.get(sock)%2} loner(s)') # equivalent to above


# # there's a built in type for this!
# from collections import Counter
# counter = Counter(sock_list)
# print(counter)

File: 3 JavaScript/solutions/sock_sorter.js
// sock_sorter.js
const sockTypes = ['ankle', 'crew', 'calf', 'thigh']
const sockList = []
const sockCounter = {}

for (let i=0; i<100; i++) {
  // randomly select sock from socktypes
  let sock = sockTypes[Math.floor(Math.random() * sockTypes.length)]
  sockList.push(sock)

  // setting default values: equivalent to dict.get() in JS 
  sockCounter[sock] =  (sockCounter[sock] || 0) + 1

  // if (sockCounter.hasOwnProperty(sock)) { // sock exists in counter
  //   sockCounter[sock] += 1 // increase count by 1
  // } else {
  //   sockCounter[sock] = 1 // set count as 1
  // }
}

for (let sock in sockCounter) {
  console.log(`${sock} has ${sockCounter[sock] % 2} loner(s).`)
}

console.log(sockList)
console.log(sockCounter)